File: src/ingestion/from-anthropic.ts
import consola from "consola"

import type {
  CanonicalRequest,
  CIFAssistantMessage,
  CIFContentPart,
  CIFMessage,
  CIFTool,
  CIFToolChoice,
  CIFUserMessage,
} from "~/cif/types"
import type {
  AnthropicAssistantContentBlock,
  AnthropicAssistantMessage,
  AnthropicMessage,
  AnthropicMessagesPayload,
  AnthropicTextBlock,
  AnthropicTool,
  AnthropicUserContentBlock,
  AnthropicUserMessage,
} from "~/routes/messages/anthropic-types"

/**
 * Convert Anthropic Messages API payload to CanonicalRequest
 */
export function parseAnthropicMessages(
  payload: AnthropicMessagesPayload,
): CanonicalRequest {
  consola.debug(
    `[Anthropic Ingestion] Parsing Anthropic payload for model: ${payload.model}`,
  )
  consola.debug(
    `[Anthropic Ingestion] Payload contains ${payload.messages.length} messages, stream: ${payload.stream ?? false}`,
  )

  const systemPrompt = translateSystemPrompt(payload.system)
  const messages = translateAnthropicMessages(payload.messages)
  const tools = translateAnthropicTools(payload.tools)

  if (systemPrompt) {
    consola.debug(
      `[Anthropic Ingestion] Extracted system prompt: ${systemPrompt.length} chars`,
    )
  }
  consola.debug(
    `[Anthropic Ingestion] Converted ${messages.length} messages to CIF format`,
  )
  if (tools) {
    consola.debug(
      `[Anthropic Ingestion] Converted ${tools.length} tools to CIF format`,
    )
  }
  if (payload.thinking?.budget_tokens) {
    consola.debug(
      `[Anthropic Ingestion] Thinking budget tokens: ${payload.thinking.budget_tokens}`,
    )
  }

  return {
    model: payload.model, // Keep original model name - adapters handle remapping
    systemPrompt,
    messages,
    tools,
    toolChoice: translateAnthropicToolChoice(payload.tool_choice),
    temperature: payload.temperature ?? undefined,
    topP: payload.top_p ?? undefined,
    maxTokens: payload.max_tokens,
    stop: payload.stop_sequences,
    stream: payload.stream ?? false,
    userId: payload.metadata?.user_id,
    extensions: {
      thinkingBudgetTokens: payload.thinking?.budget_tokens,
    },
  }
}

function translateSystemPrompt(
  system: string | Array<AnthropicTextBlock> | undefined,
): string | undefined {
  if (!system) {
    return undefined
  }

  if (typeof system === "string") {
    consola.debug(
      `[Anthropic Ingestion] System prompt is string: ${system.length} chars`,
    )
    return system
  }

  consola.debug(
    `[Anthropic Ingestion] System prompt has ${system.length} text blocks`,
  )
  // Join multiple system blocks with double newline
  return system.map((block) => block.text).join("\n\n")
}

function translateAnthropicMessages(
  anthropicMessages: Array<AnthropicMessage>,
): Array<CIFMessage> {
  consola.debug(
    `[Anthropic Ingestion] Processing ${anthropicMessages.length} Anthropic messages`,
  )

  const cifMessages = anthropicMessages.map((message) =>
    message.role === "user" ?
      translateUserMessage(message)
    : translateAssistantMessage(message),
  )

  // Log message type distribution
  const roleCount = cifMessages.reduce<Record<string, number>>((acc, msg) => {
    acc[msg.role] = (acc[msg.role] || 0) + 1
    return acc
  }, {})
  consola.debug(`[Anthropic Ingestion] Message distribution:`, roleCount)

  return cifMessages
}

function translateUserMessage(message: AnthropicUserMessage): CIFUserMessage {
  if (typeof message.content === "string") {
    return {
      role: "user",
      content: [{ type: "text", text: message.content }],
    }
  }

  return {
    role: "user",
    content: message.content.map(translateUserContentBlock),
  }
}

function translateUserContentBlock(
  block: AnthropicUserContentBlock,
): CIFContentPart {
  switch (block.type) {
    case "text": {
      return { type: "text", text: block.text }
    }
    case "image": {
      return {
        type: "image",
        mediaType: block.source.media_type,
        data: block.source.data,
      }
    }
    case "tool_result": {
      return {
        type: "tool_result",
        toolCallId: block.tool_use_id,
        toolName: "", // Will be filled in by provider adapter if needed
        content: normalizeToolResultContent(block.content),
        isError: block.is_error,
      }
    }
    default: {
      // TypeScript exhaustiveness check
      const _exhaustive: never = block
      throw new Error(
        `Unknown user content block type: ${JSON.stringify(_exhaustive)}`,
      )
    }
  }
}

function normalizeToolResultContent(
  content: string | Array<AnthropicTextBlock>,
): string {
  if (typeof content === "string") {
    return content
  }

  return content
    .map((block) => block.text)
    .filter((text) => text.length > 0)
    .join("\n\n")
}

function translateAssistantMessage(
  message: AnthropicAssistantMessage,
): CIFAssistantMessage {
  if (typeof message.content === "string") {
    return {
      role: "assistant",
      content: [{ type: "text", text: message.content }],
    }
  }

  return {
    role: "assistant",
    content: message.content.map(translateAssistantContentBlock),
  }
}

function translateAssistantContentBlock(
  block: AnthropicAssistantContentBlock,
): CIFContentPart {
  switch (block.type) {
    case "text": {
      return { type: "text", text: block.text }
    }
    case "thinking": {
      return {
        type: "thinking",
        thinking: block.thinking,
      }
    }
    case "tool_use": {
      return {
        type: "tool_call",
        toolCallId: block.id,
        toolName: block.name,
        toolArguments: block.input,
      }
    }
    default: {
      // TypeScript exhaustiveness check
      const _exhaustive: never = block
      throw new Error(
        `Unknown assistant content block type: ${JSON.stringify(_exhaustive)}`,
      )
    }
  }
}

function translateAnthropicTools(
  anthropicTools: Array<AnthropicTool> | undefined,
): Array<CIFTool> | undefined {
  if (!anthropicTools) {
    return undefined
  }

  return anthropicTools.map((tool) => ({
    name: tool.name,
    description: tool.description,
    parametersSchema: normalizeAnthropicToolInputSchema(tool.input_schema),
  }))
}

function translateAnthropicToolChoice(
  anthropicToolChoice: AnthropicMessagesPayload["tool_choice"],
): CIFToolChoice | undefined {
  if (!anthropicToolChoice) {
    return undefined
  }

  switch (anthropicToolChoice.type) {
    case "auto": {
      return "auto"
    }
    case "any": {
      return "required"
    }
    case "tool": {
      if (anthropicToolChoice.name) {
        return {
          type: "function",
          functionName: anthropicToolChoice.name,
        }
      }
      return undefined
    }
    case "none": {
      return "none"
    }
    default: {
      return undefined
    }
  }
}

// ─────────────────────────────────────────────────────────
// JSON Schema normalization (copied from non-stream-translation.ts)
// ─────────────────────────────────────────────────────────

function normalizeAnthropicToolInputSchema(
  inputSchema: Record<string, unknown>,
): Record<string, unknown> {
  const normalized = normalizeJsonSchema(inputSchema)

  if (!looksLikeJsonSchema(normalized)) {
    return {
      type: "object",
      properties: Object.fromEntries(
        Object.entries(normalized).map(([key, value]) => [
          key,
          normalizeJsonSchema(value),
        ]),
      ),
    }
  }

  return normalized
}

function normalizeJsonSchema(value: unknown): Record<string, unknown> {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    return {}
  }

  const schema = { ...(value as Record<string, unknown>) }
  const nullable = schema.nullable === true
  delete schema.nullable

  // Remove banned fields that might cause issues with providers
  const banned = new Set([
    "$schema",
    "$id",
    "patternProperties",
    "prefill",
    "enumTitles",
    "deprecated",
    "propertyNames",
    "exclusiveMinimum",
    "exclusiveMaximum",
    "const",
  ])

  for (const bannedField of banned) {
    delete schema[bannedField]
  }

  if (
    schema.properties
    && typeof schema.properties === "object"
    && !Array.isArray(schema.properties)
  ) {
    schema.properties = Object.fromEntries(
      Object.entries(schema.properties as Record<string, unknown>).map(
        ([key, child]) => [key, normalizeJsonSchema(child)],
      ),
    )
  }

  if (schema.items && typeof schema.items === "object") {
    schema.items =
      Array.isArray(schema.items) ?
        schema.items.map((item) => normalizeJsonSchema(item))
      : normalizeJsonSchema(schema.items)
  }

  for (const key of ["allOf", "anyOf", "oneOf"] as const) {
    const branch = schema[key]
    if (Array.isArray(branch)) {
      schema[key] = branch.map((item) => normalizeJsonSchema(item))
    }
  }

  if (
    schema.additionalProperties
    && typeof schema.additionalProperties === "object"
  ) {
    schema.additionalProperties = normalizeJsonSchema(
      schema.additionalProperties,
    )
  }

  if (!schema.type) {
    if (
      "properties" in schema
      || "required" in schema
      || "additionalProperties" in schema
    ) {
      schema.type = "object"
    } else if ("items" in schema) {
      schema.type = "array"
    }
  }

  if (nullable) {
    schema.type =
      typeof schema.type === "string" ? [schema.type, "null"]
      : Array.isArray(schema.type) ? [...new Set([...schema.type, "null"])]
      : ["null"]
  }

  return schema
}

function looksLikeJsonSchema(schema: Record<string, unknown>): boolean {
  const schemaKeys = new Set([
    "$defs",
    "$id",
    "$ref",
    "$schema",
    "additionalProperties",
    "allOf",
    "anyOf",
    "const",
    "default",
    "description",
    "enum",
    "examples",
    "format",
    "items",
    "maxItems",
    "maxLength",
    "maximum",
    "minItems",
    "minLength",
    "minimum",
    "multipleOf",
    "not",
    "oneOf",
    "pattern",
    "properties",
    "required",
    "title",
    "type",
  ])

  return Object.keys(schema).some((key) => schemaKeys.has(key))
}

File: internal/ingestion/from_anthropic.go
package ingestion

import (
	"encoding/json"
	"fmt"
	"omnillm/internal/cif"
	"strings"

	"github.com/rs/zerolog/log"
)

// Anthropic Messages format
type AnthropicMessage struct {
	Role    string      `json:"role"`
	Content interface{} `json:"content"` // string or array of content blocks
}

type AnthropicContentBlock struct {
	Type      string                 `json:"type"`
	Text      string                 `json:"text,omitempty"`
	Thinking  string                 `json:"thinking,omitempty"`
	Signature string                 `json:"signature,omitempty"`
	Source    *AnthropicImageSource  `json:"source,omitempty"`
	ID        string                 `json:"id,omitempty"`
	ToolUseID string                 `json:"tool_use_id,omitempty"`
	Name      string                 `json:"name,omitempty"`
	Input     map[string]interface{} `json:"input,omitempty"`
	Content   interface{}            `json:"content,omitempty"`
	IsError   *bool                  `json:"is_error,omitempty"`
}

type AnthropicImageSource struct {
	Type      string `json:"type"`
	MediaType string `json:"media_type"`
	Data      string `json:"data"`
}

type AnthropicTool struct {
	Name        string                 `json:"name"`
	Description string                 `json:"description,omitempty"`
	InputSchema map[string]interface{} `json:"input_schema"`
}

type AnthropicMessagesRequest struct {
	Model       string             `json:"model"`
	System      interface{}        `json:"system,omitempty"`
	Messages    []AnthropicMessage `json:"messages"`
	Tools       []AnthropicTool    `json:"tools,omitempty"`
	ToolChoice  interface{}        `json:"tool_choice,omitempty"`
	Temperature *float64           `json:"temperature,omitempty"`
	TopP        *float64           `json:"top_p,omitempty"`
	MaxTokens   *int               `json:"max_tokens,omitempty"`
	Stop        []string           `json:"stop_sequences,omitempty"`
	Stream      *bool              `json:"stream,omitempty"`
	Metadata    *struct {
		UserID string `json:"user_id,omitempty"`
	} `json:"metadata,omitempty"`
}

// ParseAnthropicMessages converts Anthropic messages format to CIF
func ParseAnthropicMessages(raw json.RawMessage) (*cif.CanonicalRequest, error) {
	var req AnthropicMessagesRequest
	if err := json.Unmarshal(raw, &req); err != nil {
		return nil, fmt.Errorf("failed to unmarshal Anthropic request: %w", err)
	}

	canonical := &cif.CanonicalRequest{
		Model:       req.Model,
		Temperature: req.Temperature,
		TopP:        req.TopP,
		MaxTokens:   req.MaxTokens,
		Stop:        req.Stop,
		Stream:      req.Stream != nil && *req.Stream,
	}

	if systemPrompt := normalizeAnthropicSystem(req.System); systemPrompt != nil {
		canonical.SystemPrompt = systemPrompt
	}
	if req.Metadata != nil && req.Metadata.UserID != "" {
		canonical.UserID = &req.Metadata.UserID
	}

	for _, msg := range req.Messages {
		cifMsg, err := convertAnthropicMessage(msg)
		if err != nil {
			return nil, fmt.Errorf("failed to convert message: %w", err)
		}
		canonical.Messages = append(canonical.Messages, cifMsg)
	}

	for _, tool := range req.Tools {
		var description *string
		if tool.Description != "" {
			description = &tool.Description
		}
		cifTool := cif.CIFTool{
			Name:             tool.Name,
			Description:      description,
			ParametersSchema: normalizeSchema(tool.InputSchema),
		}
		canonical.Tools = append(canonical.Tools, cifTool)
	}

	canonical.ToolChoice = normalizeAnthropicToolChoice(req.ToolChoice)

	log.Debug().
		Str("model", canonical.Model).
		Int("messages", len(canonical.Messages)).
		Int("tools", len(canonical.Tools)).
		Bool("stream", canonical.Stream).
		Msg("Converted Anthropic request to CIF")

	return canonical, nil
}

func convertAnthropicMessage(msg AnthropicMessage) (cif.CIFMessage, error) {
	contentParts, err := convertAnthropicMessageContent(msg.Content)
	if err != nil {
		return nil, err
	}

	switch msg.Role {
	case "user":
		return cif.CIFUserMessage{
			Role:    "user",
			Content: contentParts,
		}, nil

	case "assistant":
		return cif.CIFAssistantMessage{
			Role:    "assistant",
			Content: contentParts,
		}, nil

	default:
		return nil, fmt.Errorf("unknown message role: %s", msg.Role)
	}
}

func convertAnthropicMessageContent(content interface{}) ([]cif.CIFContentPart, error) {
	switch raw := content.(type) {
	case string:
		return []cif.CIFContentPart{
			cif.CIFTextPart{Type: "text", Text: raw},
		}, nil
	case []interface{}:
		contentParts := make([]cif.CIFContentPart, 0, len(raw))
		for _, blockRaw := range raw {
			blockMap, ok := blockRaw.(map[string]interface{})
			if !ok {
				return nil, fmt.Errorf("invalid Anthropic content block type: %T", blockRaw)
			}
			// Build typed block directly from the map to avoid a
			// marshal→unmarshal roundtrip for each content block.
			block := anthropicBlockFromMap(blockMap)
			cifPart, err := convertAnthropicContentBlock(block)
			if err != nil {
				return nil, err
			}
			contentParts = append(contentParts, cifPart)
		}
		return contentParts, nil
	case nil:
		return []cif.CIFContentPart{}, nil
	default:
		return nil, fmt.Errorf("unsupported Anthropic message content type: %T", content)
	}
}

// anthropicBlockFromMap extracts AnthropicContentBlock fields directly from a
// map[string]interface{} produced by the top-level json.Unmarshal call.  This
// avoids a second marshal+unmarshal cycle for every content block.
func anthropicBlockFromMap(m map[string]interface{}) AnthropicContentBlock {
	block := AnthropicContentBlock{}
	block.Type, _ = m["type"].(string)
	block.Text, _ = m["text"].(string)
	block.Thinking, _ = m["thinking"].(string)
	block.Signature, _ = m["signature"].(string)
	block.ID, _ = m["id"].(string)
	block.ToolUseID, _ = m["tool_use_id"].(string)
	block.Name, _ = m["name"].(string)
	block.Content = m["content"]
	block.Input, _ = m["input"].(map[string]interface{})

	if srcRaw, ok := m["source"].(map[string]interface{}); ok {
		src := &AnthropicImageSource{}
		src.Type, _ = srcRaw["type"].(string)
		src.MediaType, _ = srcRaw["media_type"].(string)
		src.Data, _ = srcRaw["data"].(string)
		block.Source = src
	}

	if isErr, ok := m["is_error"].(bool); ok {
		block.IsError = &isErr
	}

	return block
}

func convertAnthropicContentBlock(block AnthropicContentBlock) (cif.CIFContentPart, error) {
	switch block.Type {
	case "text":
		return cif.CIFTextPart{
			Type: "text",
			Text: block.Text,
		}, nil

	case "image":
		if block.Source != nil {
			part := cif.CIFImagePart{
				Type:      "image",
				MediaType: block.Source.MediaType,
			}
			if block.Source.Type == "base64" {
				part.Data = &block.Source.Data
			}
			return part, nil
		}
		return nil, fmt.Errorf("image block missing source")

	case "thinking":
		part := cif.CIFThinkingPart{
			Type:     "thinking",
			Thinking: block.Thinking,
		}
		if block.Signature != "" {
			signature := block.Signature
			part.Signature = &signature
		}
		return part, nil

	case "tool_use":
		toolCallID := block.ToolUseID
		if toolCallID == "" {
			toolCallID = block.ID
		}
		return cif.CIFToolCallPart{
			Type:          "tool_call",
			ToolCallID:    toolCallID,
			ToolName:      block.Name,
			ToolArguments: block.Input,
		}, nil

	case "tool_result":
		toolCallID := block.ToolUseID
		if toolCallID == "" {
			toolCallID = block.ID
		}
		return cif.CIFToolResultPart{
			Type:       "tool_result",
			ToolCallID: toolCallID,
			ToolName:   block.Name,
			Content:    normalizeAnthropicToolResultContent(block.Content),
			IsError:    block.IsError,
		}, nil

	default:
		return nil, fmt.Errorf("unknown content block type: %s", block.Type)
	}
}

func normalizeAnthropicSystem(system interface{}) *string {
	switch value := system.(type) {
	case string:
		if value == "" {
			return nil
		}
		return &value
	case []interface{}:
		parts := make([]string, 0, len(value))
		for _, rawPart := range value {
			partMap, ok := rawPart.(map[string]interface{})
			if !ok {
				continue
			}
			if partType, _ := partMap["type"].(string); partType != "text" {
				continue
			}
			if text, ok := partMap["text"].(string); ok {
				parts = append(parts, text)
			}
		}
		if len(parts) == 0 {
			return nil
		}
		result := strings.Join(parts, "\n\n")
		return &result
	default:
		return nil
	}
}

func normalizeAnthropicToolChoice(choice interface{}) interface{} {
	choiceMap, ok := choice.(map[string]interface{})
	if !ok {
		return nil
	}

	switch choiceType, _ := choiceMap["type"].(string); choiceType {
	case "auto":
		return "auto"
	case "any":
		return "required"
	case "none":
		return "none"
	case "tool":
		if name, ok := choiceMap["name"].(string); ok && name != "" {
			return map[string]interface{}{
				"type":         "function",
				"functionName": name,
			}
		}
	}

	return nil
}

func normalizeAnthropicToolResultContent(content interface{}) string {
	switch value := content.(type) {
	case string:
		return value
	case []interface{}:
		parts := make([]string, 0, len(value))
		for _, rawPart := range value {
			partMap, ok := rawPart.(map[string]interface{})
			if !ok {
				continue
			}
			if text, ok := partMap["text"].(string); ok {
				parts = append(parts, text)
			}
		}
		if len(parts) > 0 {
			return strings.Join(parts, "\n\n")
		}
	case nil:
		return ""
	}

	if jsonBytes, err := json.Marshal(content); err == nil {
		return string(jsonBytes)
	}
	return ""
}

func normalizeSchema(schema interface{}) map[string]interface{} {
	normalized, ok := normalizeSchemaValue(schema).(map[string]interface{})
	if !ok {
		return nil
	}
	return normalized
}

func normalizeSchemaValue(value interface{}) interface{} {
	switch typed := value.(type) {
	case map[string]interface{}:
		result := make(map[string]interface{}, len(typed))
		for key, item := range typed {
			if key == "$schema" || key == "patternProperties" {
				continue
			}
			result[key] = normalizeSchemaValue(item)
		}

		if nullable, ok := result["nullable"].(bool); ok && nullable {
			delete(result, "nullable")
			switch currentType := result["type"].(type) {
			case string:
				result["type"] = []interface{}{currentType, "null"}
			case []interface{}:
				hasNull := false
				for _, item := range currentType {
					if text, ok := item.(string); ok && text == "null" {
						hasNull = true
						break
					}
				}
				if !hasNull {
					result["type"] = append(currentType, "null")
				}
			}
		}

		return result
	case []interface{}:
		result := make([]interface{}, 0, len(typed))
		for _, item := range typed {
			result = append(result, normalizeSchemaValue(item))
		}
		return result
	default:
		return value
	}
}
